File: Source/TextField.m
#import "TextField.h"

#import "TextFieldTypeManager.h"
#import "TextFieldClearButton.h"

@import Hex;

static const CGFloat FieldCellLeftMargin = 10.0f;
static const CGFloat TextFieldClearButtonWidth = 30.0f;
static const CGFloat TextFieldClearButtonHeight = 20.0f;

@interface TextField () <UITextFieldDelegate>

@property (nonatomic) UIButton *customClearButton;

@property (nonatomic) UIColor *activeBackgroundColor;
@property (nonatomic) UIColor *activeBorderColor;
@property (nonatomic) UIColor *inactiveBackgroundColor;
@property (nonatomic) UIColor *inactiveBorderColor;

@property (nonatomic) UIColor *enabledBackgroundColor;
@property (nonatomic) UIColor *enabledBorderColor;
@property (nonatomic) UIColor *enabledTextColor;
@property (nonatomic) UIColor *disabledBackgroundColor;
@property (nonatomic) UIColor *disabledBorderColor;
@property (nonatomic) UIColor *disabledTextColor;

@property (nonatomic) UIColor *validBackgroundColor;
@property (nonatomic) UIColor *validBorderColor;
@property (nonatomic) UIColor *invalidBackgroundColor;
@property (nonatomic) UIColor *invalidBorderColor;
@property (nonatomic) UIColor *accessoryButtonColor;

@end

@implementation TextField

@synthesize rawText = _rawText;

#pragma mark - Initializers

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (!self) return nil;

    self.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
    self.delegate = self;

    UIView *paddingView = [[UIView alloc] initWithFrame:CGRectMake(0.0f, 0.0f, FieldCellLeftMargin, 0.0f)];
    self.leftView = paddingView;
    self.leftViewMode = UITextFieldViewModeAlways;

    [self addTarget:self action:@selector(textFieldDidUpdate:) forControlEvents:UIControlEventEditingChanged];
    [self addTarget:self action:@selector(textFieldDidReturn:) forControlEvents:UIControlEventEditingDidEndOnExit];

    self.returnKeyType = UIReturnKeyDone;
    self.rightViewMode = UITextFieldViewModeWhileEditing;

    return self;
}

#pragma mark - Getters

- (UIButton *)customClearButton {
    if (!_customClearButton) {
        UIImage *clearImage = [TextFieldClearButton imageForSize:CGSizeMake(18, 18) andButtonType:TextFieldButtonTypeClear color:self.accessoryButtonColor];
        _customClearButton = [UIButton buttonWithType:UIButtonTypeCustom];
        [_customClearButton setImage:clearImage forState:UIControlStateNormal];
        [_customClearButton addTarget:self action:@selector(clearButtonAction) forControlEvents:UIControlEventTouchUpInside];
        _customClearButton.frame = CGRectMake(0.0f, 0.0f, TextFieldClearButtonWidth, TextFieldClearButtonHeight);
    }

    return _customClearButton;
}

#pragma mark - Setters

- (NSRange)currentRange {
    NSInteger startOffset = [self offsetFromPosition:self.beginningOfDocument
                                          toPosition:self.selectedTextRange.start];
    NSInteger endOffset = [self offsetFromPosition:self.beginningOfDocument
                                        toPosition:self.selectedTextRange.end];
    NSRange range = NSMakeRange(startOffset, endOffset-startOffset);

    return range;
}

- (void)setText:(NSString *)text {
    UITextRange *textRange = self.selectedTextRange;
    NSString *newRawText = [self.formatter formatString:text
                                                reverse:YES];
    NSRange range = [self currentRange];

    BOOL didAddText  = (newRawText.length > self.rawText.length);
    BOOL didFormat   = (text.length > super.text.length);
    BOOL cursorAtEnd = (newRawText.length == range.location);

    if ((didAddText && didFormat) || (didAddText && cursorAtEnd)) {
        self.selectedTextRange = textRange;
        [super setText:text];
    } else {
        [super setText:text];
        self.selectedTextRange = textRange;
    }
}

- (void)setRawText:(NSString *)rawText {
    BOOL shouldFormat = (self.formatter && (rawText.length >= _rawText.length ||
                                            ![rawText isEqualToString:_rawText]));

    if (shouldFormat) {
        self.text = [self.formatter formatString:rawText reverse:NO];
    } else {
        self.text = rawText;
    }

    _rawText = rawText;
}

- (void)setInputTypeString:(NSString *)inputTypeString {
    _inputTypeString = inputTypeString;

    TextFieldInputType inputType;
    if ([inputTypeString isEqualToString:@"name"]) {
        inputType = TextFieldInputTypeName;
    } else if ([inputTypeString isEqualToString:@"username"]) {
        inputType = TextFieldInputTypeUsername;
    } else if ([inputTypeString isEqualToString:@"phone"]) {
        inputType = TextFieldInputTypePhoneNumber;
    } else if ([inputTypeString isEqualToString:@"number"]) {
        inputType = TextFieldInputTypeNumber;
    } else if ([inputTypeString isEqualToString:@"float"]) {
        inputType = TextFieldInputTypeFloat;
    } else if ([inputTypeString isEqualToString:@"address"]) {
        inputType = TextFieldInputTypeAddress;
    } else if ([inputTypeString isEqualToString:@"email"]) {
        inputType = TextFieldInputTypeEmail;
    } else if ([inputTypeString isEqualToString:@"text"]) {
        inputType = TextFieldInputTypeDefault;
    } else if ([inputTypeString isEqualToString:@"password"]) {
        inputType = TextFieldInputTypePassword;
    } else if ([inputTypeString isEqualToString:@"count"]) {
        inputType = TextFieldInputTypeCount;
    } else if (!inputTypeString.length) {
        inputType = TextFieldInputTypeDefault;
    } else {
        inputType = TextFieldInputTypeUnknown;
    }

    self.inputType = inputType;
}

- (void)setInputType:(TextFieldInputType)inputType {
    _inputType = inputType;

    TextFieldTypeManager *typeManager = [TextFieldTypeManager new];
    [typeManager setUpType:inputType forTextField:self];
}

#pragma mark - Getters

- (NSString *)rawText {
    if (self.formatter) {
        return [self.formatter formatString:_rawText reverse:YES];
    }

    return _rawText;
}

#pragma mark - Public

- (BOOL)validate {
    BOOL isValid = [self.inputValidator validateString:self.text];
    self.valid = isValid;
    return isValid;
}

#pragma mark - UITextFieldDelegate

- (void)textFieldDidBeginEditing:(TextField *)textField {
    [self updateActive:YES];
}

- (void)textFieldDidEndEditing:(UITextField *)textField {
    [self updateActive:NO];
}

- (BOOL)textField:(UITextField *)textField shouldChangeCharactersInRange:(NSRange)range replacementString:(NSString *)string {
    if (!string || [string isEqualToString:@"\n"]) {
        return YES;
    }

    BOOL validator = (self.inputValidator &&
                      [self.inputValidator respondsToSelector:@selector(validateReplacementString:withText:withRange:)]);

    if (validator) {
        return [self.inputValidator validateReplacementString:string
                                                     withText:self.text withRange:range];
    }

    return YES;
}

#pragma mark - Notifications

- (void)textFieldDidUpdate:(UITextField *)textField {
    if (!self.isValid) {
        self.valid = YES;
    }

    self.rawText = self.text;

    if ([self.textFieldDelegate respondsToSelector:@selector(textFormField:didUpdateWithText:)]) {
        [self.textFieldDelegate textFormField:self
                            didUpdateWithText:self.rawText];
    }
}

- (void)textFieldDidReturn:(UITextField *)textField {
    if ([self.textFieldDelegate respondsToSelector:@selector(textFormFieldDidReturn:)]) {
        [self.textFieldDelegate textFormFieldDidReturn:self];
    }
}

#pragma mark - Actions

- (void)clearButtonAction {
    self.rawText = nil;

    if ([self.textFieldDelegate respondsToSelector:@selector(textFormField:didUpdateWithText:)]) {
        [self.textFieldDelegate textFormField:self
                            didUpdateWithText:self.rawText];
    }
}

#pragma mark - Appearance

- (void)setEnabled:(BOOL)enabled {
    [super setEnabled:enabled];

    [self updateEnabled:enabled];
}

- (void)setValid:(BOOL)valid {
    _valid = valid;

    if (!self.isEnabled) return;

    [self updateValid:valid];
}

- (void)updateActive:(BOOL)active {
    self.rightView = self.customClearButton;

    if (active) {
        self.backgroundColor = self.activeBackgroundColor;
        self.layer.backgroundColor = self.activeBackgroundColor.CGColor;
        self.layer.borderColor = self.activeBorderColor.CGColor;
    } else {
        self.backgroundColor = self.inactiveBackgroundColor;
        self.layer.backgroundColor = self.inactiveBackgroundColor.CGColor;
        self.layer.borderColor = self.inactiveBorderColor.CGColor;
    }
}

- (void)updateEnabled:(BOOL)enabled {
    if (enabled) {
        self.backgroundColor = self.enabledBackgroundColor;
        self.layer.borderColor = self.enabledBorderColor.CGColor;
        self.layer.backgroundColor = self.enabledBackgroundColor.CGColor;
        self.textColor = self.enabledTextColor;
    } else {
        self.backgroundColor = self.disabledBackgroundColor;
        self.layer.borderColor = self.disabledBorderColor.CGColor;
        self.layer.backgroundColor = self.disabledBackgroundColor.CGColor;
        self.textColor = self.disabledTextColor;
    }
}

- (void)updateValid:(BOOL)valid {
    if (valid) {
        self.backgroundColor = self.validBackgroundColor;
        self.layer.borderColor = self.validBorderColor.CGColor;
    } else {
        self.backgroundColor = self.invalidBackgroundColor;
        self.layer.borderColor = self.invalidBorderColor.CGColor;
    }
}

#pragma mark - Styling

- (void)setCustomFont:(UIFont *)font {
    if (font) {
        self.font = font;
    }
}

- (void)setBorderWidth:(CGFloat)borderWidth {
    self.layer.borderWidth = borderWidth;
}

- (void)setBorderColor:(UIColor *)borderColor {
    if (borderColor) {
        self.layer.borderColor = borderColor.CGColor;
    }
}

- (void)setCornerRadius:(CGFloat)cornerRadius {
    self.layer.cornerRadius = cornerRadius;
}

- (void)setActiveBackgroundColor:(UIColor *)color {
    _activeBackgroundColor = color;
}

- (void)setActiveBorderColor:(UIColor *)color {
    _activeBorderColor = color;
}

- (void)setInactiveBackgroundColor:(UIColor *)color {
    _inactiveBackgroundColor = color;
}

- (void)setInactiveBorderColor:(UIColor *)color {
    _inactiveBorderColor = color;
}

- (void)setEnabledBackgroundColor:(UIColor *)color {
    _enabledBackgroundColor = color;

    [self updateEnabled:self.enabled];
}

- (void)setEnabledBorderColor:(UIColor *)color {
    _enabledBorderColor = color;

    [self updateEnabled:self.enabled];
}

- (void)setEnabledTextColor:(UIColor *)color {
    _enabledTextColor = color;

    [self updateEnabled:self.enabled];
}

- (void)setDisabledBackgroundColor:(UIColor *)color {
    _disabledBackgroundColor = color;

    [self updateEnabled:self.enabled];
}

- (void)setDisabledBorderColor:(UIColor *)color {
    _disabledBorderColor = color;

    [self updateEnabled:self.enabled];
}

- (void)setDisabledTextColor:(UIColor *)color {
    _disabledTextColor = color;

    [self updateEnabled:self.enabled];
}

- (void)setValidBackgroundColor:(UIColor *)color {
    _validBackgroundColor = color;
}

- (void)setValidBorderColor:(UIColor *)color {
    _validBorderColor = color;
}

- (void)setInvalidBackgroundColor:(UIColor *)color {
    _invalidBackgroundColor = color;
}

- (void)setInvalidBorderColor:(UIColor *)color {
    _invalidBorderColor = color;
}

- (void)setAccessoryButtonColor:(UIColor *)color {
    _accessoryButtonColor = color;
}

@end

File: Source/TextField.swift
 import UIKit
import InputValidator
import Formatter

public enum TextFieldInputType {
    case Default, Name, Username, PhoneNumber, Integer, Decimal, Address, Email, Password, Unknown
}

public protocol TextFieldDelegate: class {
    func didBeginEditing(textField: TextField)
    func didEndEditing(textField: TextField)
    func didUpdateWithText(text: String?, textField: TextField)
    func didReturn(textField: TextField)
}

public class TextField: UITextField, UITextFieldDelegate {
    dynamic public var borderWidth: CGFloat = 0 { didSet { self.layer.borderWidth = borderWidth } }
    dynamic public var cornerRadius: CGFloat = 0 { didSet { self.layer.cornerRadius = cornerRadius } }
    dynamic public var accessoryButtonColor: UIColor = UIColor.redColor()

    dynamic public var enabledBackgroundColor: UIColor = UIColor.redColor() { didSet { self.updateEnabled(self.enabled) } }
    dynamic public var enabledBorderColor: UIColor = UIColor.redColor() { didSet { self.updateEnabled(self.enabled) } }
    dynamic public var enabledTextColor: UIColor = UIColor.redColor() { didSet { self.updateEnabled(self.enabled) } }

    dynamic public var validBackgroundColor: UIColor = UIColor.redColor()
    dynamic public var validBorderColor: UIColor = UIColor.redColor()
    dynamic public var validTextColor: UIColor = UIColor.redColor()

    dynamic public var activeBackgroundColor: UIColor = UIColor.redColor()
    dynamic public var activeBorderColor: UIColor = UIColor.redColor()
    dynamic public var activeTextColor: UIColor = UIColor.redColor()

    dynamic public var inactiveBackgroundColor: UIColor = UIColor.redColor()
    dynamic public var inactiveBorderColor: UIColor = UIColor.redColor()
    dynamic public var inactiveTextColor: UIColor = UIColor.redColor()

    dynamic public var disabledBackgroundColor: UIColor = UIColor.redColor() { didSet { self.updateEnabled(self.enabled) } }
    dynamic public var disabledBorderColor: UIColor = UIColor.redColor() { didSet { self.updateEnabled(self.enabled) } }
    dynamic public var disabledTextColor: UIColor = UIColor.redColor() { didSet { self.updateEnabled(self.enabled) } }

    dynamic public var invalidBackgroundColor: UIColor = UIColor.redColor()
    dynamic public var invalidBorderColor: UIColor = UIColor.redColor()
    dynamic public var invalidTextColor: UIColor = UIColor.redColor()

    public var inputValidator: Validatable?
    public var formatter: Formattable?
    public weak var textFieldDelegate: TextFieldDelegate?

    static let LeftMargin = 10.0
    static let AccessoryButtonWidth = 30.0
    static let AccessoryButtonHeight = 20.0

    override public init(frame: CGRect) {
        self.inputType = .Default

        super.init(frame: frame)

        self.delegate = self

        let paddingView = UIView(frame: CGRect(x: 0, y: 0, width: TextField.LeftMargin, height: 0))
        self.leftView = paddingView
        self.leftViewMode = .Always

        self.addTarget(self, action: "textFieldDidUpdate:", forControlEvents: .EditingChanged)
        self.addTarget(self, action: "textFieldDidReturn:", forControlEvents: .EditingDidEndOnExit)

        self.returnKeyType = .Done
        self.rightViewMode = .WhileEditing
        self.backgroundColor = UIColor.clearColor()
    }

    required public init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    lazy var customClearButton: UIButton = {
        let image = TextFieldClearButton.imageForSize(CGSize(width: 18, height: 18), color: self.accessoryButtonColor)
        let button = UIButton(type: .Custom)
        button.setImage(image, forState: .Normal)
        button.addTarget(self, action: "clearButtonAction", forControlEvents: .TouchUpInside)
        button.frame = CGRect(x: 0, y: 0, width: TextField.AccessoryButtonWidth, height: TextField.AccessoryButtonHeight)

        return button
    }()

    override public var enabled: Bool {
        didSet {
            self.updateEnabled(self.enabled)
        }
    }

    public var valid: Bool = true {
        didSet {
            if self.enabled {
                self.updateValid(self.valid)
            }
        }
    }

    public var inputType: TextFieldInputType {
        didSet {
            self.updateInputType(inputType)
        }
    }

    func updateText(newValue: String?) {
        let text = newValue ?? ""

        if self.formatter != nil {
            let textRange = self.selectedTextRange
            let newRawText = self.formatter!.formatString(text, reverse: false)

            let didAddText = (newRawText.characters.count > (self.text ?? "").characters.count)
            let didFormat = (newRawText.characters.count > (self.text ?? "").characters.count)
            let cursorAtStart = (self.selectedTextRange!.start == self.positionFromPosition(self.beginningOfDocument, offset: 1))
            if (didAddText && cursorAtStart) {
                self.text = newRawText
                self.selectedTextRange = textRange
            } else if (didAddText && didFormat) {
                super.text = newRawText
            } else {
                super.text = newRawText
                self.selectedTextRange = textRange
            }
        } else {
            self.text = text
        }
    }

    // MARK: Public

    func updateActive(active: Bool) {
        self.rightView = self.customClearButton

        if active {
            self.layer.backgroundColor = self.activeBackgroundColor.CGColor
            self.layer.borderColor = self.activeBorderColor.CGColor
            self.textColor = self.activeTextColor
        } else {
            self.layer.backgroundColor = self.inactiveBackgroundColor.CGColor
            self.layer.borderColor = self.inactiveBorderColor.CGColor
            self.textColor = self.inactiveTextColor
        }
    }

    func updateEnabled(enabled: Bool) {
        if enabled {
            self.layer.borderColor = self.enabledBorderColor.CGColor
            self.layer.backgroundColor = self.enabledBackgroundColor.CGColor
            self.textColor = self.enabledTextColor
        } else {
            self.layer.borderColor = self.disabledBorderColor.CGColor
            self.layer.backgroundColor = self.disabledBackgroundColor.CGColor
            self.textColor = self.disabledTextColor
        }
    }

    func updateValid(valid: Bool) {
        if valid {
            self.layer.backgroundColor = self.validBackgroundColor.CGColor
            self.layer.borderColor = self.validBorderColor.CGColor
            self.textColor = self.validTextColor
        } else {
            self.layer.backgroundColor = self.invalidBackgroundColor.CGColor
            self.layer.borderColor = self.invalidBorderColor.CGColor
            self.textColor = self.invalidTextColor
        }
    }

    public func validate() -> Bool {
        var isValid = true
        if let inputValidator = self.inputValidator {
            isValid = inputValidator.validateString(self.text ?? "")
        }

        self.valid = isValid

        return isValid
    }

    // MARK: UITextFieldDelegate

    public func textFieldDidBeginEditing(textField: UITextField) {
        self.rightView = self.customClearButton

        self.updateActive(true)
    }

    public func textFieldDidEndEditing(textField: UITextField) {
        self.updateActive(false)
    }

    public func textField(textField: UITextField, shouldChangeCharactersInRange range: NSRange, replacementString string: String) -> Bool {
        if string == "\n" {
            return true
        }

        var valid = true
        if self.inputValidator != nil {
            valid = self.inputValidator!.validateReplacementString(string, usingFullString: self.text, inRange: range, exhaustive: false)
        }

        return valid
    }

    // MARK: Notification

    func textFieldDidUpdate(textField: TextField) {
        self.updateText(self.text)

        if self.valid == false {
            self.valid = true
        }

        self.textFieldDelegate?.didUpdateWithText(self.text, textField: self)
    }

    func textFieldDidReturn(textField: TextField) {
        self.textFieldDelegate?.didReturn(self)
    }

    // MARK: Actions

    func clearButtonAction() {
        self.text = nil

        self.textFieldDelegate?.didUpdateWithText(self.text, textField: self)
    }
}
